### PR TITLE
@zephraph => [WIP] Webpack 4 Upgrade 

### PR DESCRIPTION
**Pull request by damassi**
#### WIP 

I started working on this a few months back but noticed that

- there was some performance degradation around compile-time (even though its advertised as being much faster) and 
- that react-hotloading wasn't quite working. 

I've verified that hot-loading has been fixed in another project, so need to update here and also need to do a little bit of cleanup and pull in latest `#master`, but wanted to push this up so that you could give this a look. 

Notes: 
- Out of memory / memory leak issues: https://github.com/webpack/webpack/issues/1914#issuecomment-416727763

### CODE REVIEW

**Comment by zephraph** [comment]
I'll dig into this. Sorry, it dropped off my radar.

**Comment by damassi** [comment]
Oh this actually isn’t ready, I just wanted to get everything as close to done before I left! There are a bunch of issues with dist bundle sizes and config.

**Comment by zephraph** [comment]
@damassi what was left w/ this?

**Comment by damassi** [comment]
@zephraph - the last remaining issue is the chunk bundle sizes which are huge. Would you mind pulling down this branch and giving it a run and seeing if anything jumps out? Otherwise I think its just about ready to go.

**Comment by zephraph** [comment]
Cool, yeah, I'll take a look.

File: webpack.config.js [hunk 1]
```diff
@@ -2,7 +2,7 @@ const ForkTsCheckerNotifierWebpackPlugin = require('fork-ts-checker-notifier-web
 const ForkTsCheckerWebpackPlugin = require('fork-ts-checker-webpack-plugin')
 const FriendlyErrorsWebpackPlugin = require('friendly-errors-webpack-plugin')
 const ProgressBarPlugin = require('progress-bar-webpack-plugin')
-const UglifyJsPlugin = require('uglifyjs-webpack-plugin')
+// const UglifyJsPlugin = require('uglifyjs-webpack-plugin')
```
Comment (damassi) [inline webpack.config.js]: TODO: Remove. No longer need this since we can now use `mode: production`.

**Comment by zephraph** [comment]
Nope, nope, nope. 

I spent some time digging into this and I ran up against a ton of memory issues. Bundles are a nightmare. Haa. One day, but not today.

**Comment by damassi** [comment]
I wonder what changed 🤔 I remember fixing this before but then updated all the deps to latest and the size issue came back. I'll look again tonight.

**Comment by zephraph** [comment]
So I removed the optimization settings in favor of Webpack's defaults. It's supposed to have some good rules around optimizations, but given the sheer number and size of our dependencies it doesn't produce an ideal output. The biggest issue I hit is that I'd get continual out of heap errors when building. A lot of it seems to spawn from uglify eating a ton of memory. After disabling minification entirely I was able to get it to generate sometimes, but there's still something causing it to choke up.

**Comment by damassi** [comment]
@zephraph - here's what it was initially set at, I wonder what would happen if this was put back: https://github.com/artsy/force/pull/2587/commits/04f76d8722bf82337f1cbc7df476c92e3c585ef9#diff-11e9f7f953edc64ba14b0cc350ae7b9dR31

**Comment by zephraph** [comment]
Old friend, alive again.

File: webpack.config.js [hunk 2]
```diff
@@ -89,9 +99,11 @@ const config = {
           },
         ],
       },
+      // ESM support. See: https://github.com/apollographql/react-apollo/issues/1737#issuecomment-371178602
       {
-        test: /\.json$/,
-        loader: "json-loader",
+        type: "javascript/auto",
+        test: /\.mjs$/,
+        use: [],
```
Comment (zephraph) [inline webpack.config.js]: This is interesting... I've never seen this being required before 🤔

Comment (damassi) [reply webpack.config.js]: Yah I needed it a while back but i'm curious if I still do...

**Comment by damassi** [comment]
(I'm actually starting this again from scratch, see where things shake out)

**Comment by zephraph** [comment]
Godspeed. More than happy to pair on it sometime.

**Comment by damassi** [comment]
Closing in favor of https://github.com/artsy/force/pull/2985

**Pull request closed without merging**